docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+    expected_local_ten = torch.arange(start=num_elems_per_rank * rank, end=num_elems_per_rank * (rank + 1))
+    assert torch.all(loaded_state_dict['weight'] == expected_local_ten)
+
+    # With torch.save and torch.load, we would have to load all files that contain
+    # parts of the desired tensor in new configuration and concatenate appropriate fragments.
+    # For some distributed checkpoint backends this is actually what happens underneath.
+
+
+Supported entities
+==================
+The distributed checkpointing library supports saving and loading of different objects in different configurations.
+
+A sharded state dict is a (possibly nested) Python dictionary or list with the following elements:
+
+1. ShardedBase
+    a. ShardedTensor
+    #. ShardedObject
+    #. ShardedTensorFactory
+#. LocalNonpersitentObject
+#. Arbitrary object
+
+
+ShardedBase
+-----------
+Base class for expressing any kind of sharding.
+Each sharded entity must be uniquely identified by its `key`, carry some `data` to be saved or loaded and define `replica_id` which helps identify data redundancy.
+
+Note that the `key` doesn't have to (and usually doesn't) correspond to the key in the state dict.
+The key in the state dict is ephemeral, while the `ShardedTensor.key` is used to identify the tensor in the checkpoint.
+
+Example:
+
+.. code-block:: python
+    import torch
+
+    from megatron.core import dist_checkpointing
+
+    # Checkpoint saved with some key in the state dict that is eventually ignored
+    model = ...
+    ckpt_dir = ...
+    sharded_state_dict = {
+        'ignored': dist_checkpointing.ShardedTensor('tensor-A', ...)
+    }
+    dist_checkpointing.save(sharded_state_dict, ckpt_dir)
+
+    # During loading, all that matters is the ShardedTensor.key.
+    sharded_state_dict = {
+        'different-key': dist_checkpointing.ShardedTensor('tensor-A', ...)
+    }
+    loaded_state_dict = dist_checkpointing.load(sharded_state_dict, ckpt_dir)
+    assert 'ignored' not in loaded_state_dict
+    assert 'tensor-A' not in loaded_state_dict
+    assert isinstance(loaded_state_dict['different-key'], torch.Tensor)
+
+    # The key in the state dict is important only from the subsequent `model.load_state_dict`
+    # that usually happens after `dist_checkpointing.load` - the state dict must have
+    # the structure and keys corresponding to the model structure and submodule names.
+    model.load_state_dict(loaded_state_dict)
+
+ShardedTensor
+-------------
+It's the primary use case of distributed checkpointing - tensors sharding.
+Allows to define how PyTorch tensors are sharded across the workload.
+See `Tensors transformations`_ section for more details on ShardedTensors.
+
+ShardedObject
+-------------
+Sometimes there is a need to save arbitrary objects across the ranks.
+ShardedObject allows to structure those objects into arrays of objects with a fixed `global_shape` and save/load parts of the arrays on specific ranks.
```

Comment:
The ShardedObject serves as a solution for saving arbitrary objects across ranks. It allows users to structure these objects into arrays with a fixed global_shape. Users can then selectively save or load parts of these arrays on specific ranks.